tools/db/migrate-external-ids/src/Work.hs

```diff
@@ -0,0 +1,159 @@
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Work where
+
+import Cassandra
+import qualified Cassandra as C
+import qualified Cassandra.Settings as C
+import Control.Lens
+import Data.Conduit
+import qualified Data.Conduit.Combinators as CC
+import Data.Conduit.Internal (zipSources)
+import qualified Data.Conduit.List as CL
+import Data.Functor.Identity
+import Data.Id
+import qualified Data.Map.Strict as Map
+import Data.Misc
+import Database.CQL.Protocol (Tuple)
+import Imports
+import Options
+import Options.Applicative
+import System.Logger (Logger)
+import qualified System.Logger as Log
+import UnliftIO.Async (pooledMapConcurrentlyN)
+import Wire.API.Team.Feature
+
+data Env = Env
+  { envBrig :: ClientState,
+    envSpar :: ClientState,
+    envLogger :: Logger,
+    envPageSize :: Int32,
+    envDebug :: Debug,
+    envDryDrun :: DryRun,
+    envFailCount :: IORef Int
+  }
+
+main :: IO ()
+main = do
+  s <- execParser (info (helper <*> settingsParser) desc)
+  lgr <- initLogger s
+  brig <- initCas (s ^. setCasBrig) lgr
+  spar <- initCas (s ^. setCasSpar) lgr
+  failCount <- newIORef 0
+  let env = Env brig spar lgr (s ^. setPageSize) (s ^. setDebug) (s ^. setDryRun) failCount
+  runCommand env
+  where
+    desc =
+      header "migrate-external-ids"
+        <> progDesc "Migrate from spar."
+        <> fullDesc
+    initLogger s =
+      Log.new
+        . Log.setOutput Log.StdOut
+        . Log.setFormat Nothing
+        . Log.setBufSize 0
+        . Log.setLogLevel
+          ( case s ^. setDebug of
+              Debug -> Log.Debug
+              _ -> Log.Warn
+          )
+        $ Log.defSettings
+    initCas cas l =
+      C.init
+        . C.setLogger (C.mkLogger l)
+        . C.setContacts (cas ^. cHosts) []
+        . C.setPortNumber (fromIntegral $ cas ^. cPort)
+        . C.setKeyspace (cas ^. cKeyspace)
+        . C.setProtocolVersion C.V4
+        $ C.defSettings
+
+runCommand :: Env -> IO ()
+runCommand env@Env {..} = do
+  runConduit $
+    zipSources
+      (CL.sourceList [(1 :: Int32) ..])
+      (readLegacyExternalIds env)
+      .| CC.mapM (resolveTeam env)
+      .| sinkMain env
+
+  failCount <- readIORef envFailCount
+  when (failCount > 0) $
+    Log.warn envLogger (Log.msg @String (show failCount <> " external ids have *NOT* been migrated.\nAn external id fails to me migrated if the mapped user doesn't exist or doesn't have a team."))
```

Comment:
```suggestion
    Log.warn envLogger (Log.msg @String (show failCount <> " external ids have *NOT* been migrated.\nAn external id fails to be migrated if the mapped user doesn't exist or doesn't have a team."))
```

Reply:
`failures :: [LegacyExternalId]` containing all failed externalIds would be more helpful.  (I'm pretty sure those fit into memory, too.)

On second thought: all the information is contained in stdout/stderr, so we can just capture that.  Furthermore, this script is idempotent, so even if we forget to capture, we can just run it again.

So this looks good to me!